### ISSUE TITLE
Potentially misleading error when holding a reference into a guard across `await` point

### ISSUE THREAD

**Issue by CeNiEi** [rust-lang/rust#136349]
I tried this code:

```rust

use parking_lot;
use tokio; // 1.43.0 // 0.12.3

use std::collections::HashMap;

struct Temp {
    temp: parking_lot::RwLock<HashMap<i32, Vec<i32>>>,
    timer: tokio::time::Interval,
}

impl Temp {
    fn poll(mut self) {
        tokio::task::spawn(async move {
            let mut _guard = self.temp.write();
            let _ = _guard.get_mut(&0);

            std::mem::drop(_guard);
            self.timer.tick().await;
        });
    }
}

fn main() {}

```

Since we are already dropping the `_guard`, before calling await. This future should be `Send` and should compile successfully.

But for some reason the compiler complains that 
```
   Compiling playground v0.0.1 (/playground)
error: future cannot be sent between threads safely
   --> src/main.rs:13:9
    |
13  | /         tokio::task::spawn(async move {
14  | |             let mut _guard = self.temp.write();
15  | |             let _ = _guard.get_mut(&0);
...   |
18  | |             self.timer.tick().await;
19  | |         });
    | |__________^ future created by async block is not `Send`
    |
    = help: within `{async block@src/main.rs:13:28: 13:38}`, the trait `Send` is not implemented for `*mut ()`
note: future is not `Send` as this value is used across an await
   --> src/main.rs:18:31
    |
14  |             let mut _guard = self.temp.write();
    |                 ---------- has type `parking_lot::lock_api::RwLockWriteGuard<'_, parking_lot::RawRwLock, HashMap<i32, Vec<i32>>>` which is not `Send`
...
18  |             self.timer.tick().await;
    |                               ^^^^^ await occurs here, with `mut _guard` maybe used later
note: required by a bound in `tokio::spawn`
   --> /playground/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.43.0/src/task/spawn.rs:168:21
    |
166 |     pub fn spawn<F>(future: F) -> JoinHandle<F::Output>
    |            ----- required by a bound in this function
167 |     where
168 |         F: Future + Send + 'static,
    |                     ^^^^ required by this bound in `spawn`

```

The issue is probably caused the `let _ = _guard.get_mut(&0);` statement. 
I guess a more reasonable error could be that bound variable is holding a ref into the guard, which can be used later (I am not sure if that is infact the case). 

### Meta

Happens on Stable 1.84.1 and Nightly 1.86.0 as well
Playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e5a6b1f10131890af79c6c6ab4e991e1
</p>
</details>

**Comment by theemathas** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/63768

Can be worked around by adding a block around the scope of the guard.

```rust
        tokio::task::spawn(async move {
            {
                let mut _guard = self.temp.write();
                let _ = _guard.get_mut(&0);
            }
            self.timer.tick().await;
        });
```
